verification/src/solidity/compiler_fetcher.rs

```diff
@@ -0,0 +1,326 @@
+    type Error = url::ParseError;
+
+    fn try_from(
+        (compiler_info, download_url): (json::CompilerInfo, &Url),
+    ) -> Result<Self, Self::Error> {
+        let url = match compiler_info.path {
+            json::DownloadPath::Url(url) => url,
+            // download_url ends with `.../list.json` but join() will replace this with `filename`
+            json::DownloadPath::Filename(filename) => download_url.join(&filename)?,
+        };
+        Ok(Self {
+            url,
+            sha256: compiler_info.sha256,
+        })
+    }
+}
+
+#[derive(Default)]
+pub struct CompilerFetcher {
+    releases: Releases,
+    folder: PathBuf,
+}
+
+impl CompilerFetcher {
+    pub async fn new(releases: Releases, folder: PathBuf) -> Self {
+        Self { releases, folder }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum FetchError {
+    #[error("version {0} not found")]
+    NotFound(CompilerVersion),
+    #[error("couldn't fetch the file: {0}")]
+    Fetch(reqwest::Error),
+    #[error("couldn't create file: {0}")]
+    File(std::io::Error),
+    #[error("tokio sheduling error: {0}")]
+    Shedule(tokio::task::JoinError),
+}
+
+#[cfg(target_family = "unix")]
+fn create_executable(path: &Path) -> Result<File, std::io::Error> {
+    OpenOptions::new()
+        .create(true)
+        .write(true)
+        .truncate(true)
+        .mode(0o777)
+        .open(path)
+}
+
+#[async_trait]
+impl Fetcher for CompilerFetcher {
+    type Error = FetchError;
+    async fn fetch(&self, ver: &CompilerVersion) -> Result<PathBuf, Self::Error> {
+        let compiler_info = self
+            .releases
+            .0
+            .get(ver)
+            .ok_or_else(|| FetchError::NotFound(ver.clone()))?;
+
+        let response = reqwest::get(compiler_info.url.clone())
+            .await
+            .map_err(FetchError::Fetch)?;
+        let folder = self.folder.join(ver.to_string());
+        let file = folder.join("solc");
+        let bytes = response.bytes().await.map_err(FetchError::Fetch)?;
+        {
+            let file = file.clone();
+            tokio::task::spawn_blocking(move || -> Result<(), Self::Error> {
+                std::fs::create_dir_all(&folder).map_err(FetchError::File)?;
+                std::fs::remove_file(file.as_path())
+                    .or_else(|e| {
+                        if e.kind() == ErrorKind::NotFound {
+                            Ok(())
+                        } else {
+                            Err(e)
+                        }
+                    })
+                    .map_err(FetchError::File)?;
+                let mut file = create_executable(file.as_path()).map_err(FetchError::File)?;
+                std::io::copy(&mut bytes.as_ref(), &mut file).map_err(FetchError::File)?;
+                Ok(())
+            })
+            .await
+            .map_err(FetchError::Shedule)??;
+        }
+
+        Ok(file)
+    }
+}
+
+mod json {
+    use super::*;
```

Comment:
Nit: what if we import only those structures that are required (`H256`, `Url`, `CompilerVersion`, and `Deserialize`)?